src/bun.js/node/types.zig

```diff
@@ -1835,548 +1849,2876 @@ pub const Emitter = struct {
+                if (maybeTmpBuf3 != null) allocator.free(maybeTmpBuf3.?);
+                return error.OutOfMemory;
+            }
+            tmpBuf1 = maybeTmpBuf1.?;
+            tmpBuf2 = maybeTmpBuf2.?;
+            tmpBuf3 = maybeTmpBuf3.?;
+            return maybeBuf.?;
+        }
+        return allocator.alloc(u8, MAX_PATH_BYTES) catch unreachable;
+    }
 
-        const base_slice = path.slice();
-        var out: []const u8 = base_slice;
+    /// Taken from Zig 0.11.0 zig/lib/std/process.zig
+    /// https://github.com/ziglang/zig/blob/776cd673f206099012d789fd5d05d49dd72b9faa/lib/std/process.zig#L363
+    ///
+    /// Caller must free returned memory.
+    fn getEnvVarOwnedW(allocator: std.mem.Allocator, key_w: []const u8) std.process.GetEnvVarOwnedError![]u8 {
+        const result_w = blk: {
+            break :blk std.os.getenvW(key_w) orelse return error.EnvironmentVariableNotFound;
+        };
+        return std.unicode.utf16leToUtf8Alloc(allocator, result_w) catch |err| switch (err) {
+            error.DanglingSurrogateHalf => return error.InvalidUtf8,
+            error.ExpectedSecondSurrogateHalf => return error.InvalidUtf8,
+            error.UnexpectedSecondSurrogateHalf => return error.InvalidUtf8,
+            else => |e| return e,
+        };
+    }
 
-        if (!isWindows) {
-            out = std.fs.path.basenamePosix(base_slice);
-        } else {
-            out = std.fs.path.basenameWindows(base_slice);
+    /// Based on Node v21.6.1 private helper posixCwd:
+    /// https://github.com/nodejs/node/blob/6ae20aa63de78294b18d5015481485b7cd8fbb60/lib/path.js#L1074
+    inline fn posixCwd(buf: []u8) []const u8 {
+        const cwd = @This().getCwd(buf);
+        const len = cwd.len;
+        if (len == 0) {
+            return "";
         }
-        const ext = extname_.slice();
+        if (comptime Environment.isWindows) {
+            // Converts Windows' backslash path separators to POSIX forward slashes
+            // and truncates any drive indicator
+
+            // Translated from the following JS code:
+            //   const cwd = StringPrototypeReplace(process.cwd(), regexp, '/');
+            for (0.. len) |i| {
+                if (cwd[i] == CHAR_BACKWARD_SLASH) {
+                    buf[i] = CHAR_FORWARD_SLASH;
+                } else {
+                    buf[i] = cwd[i];
+                }
+            }
+            var normalizedCwd = buf[0 ..len];
 
-        if ((ext.len != out.len or out.len == base_slice.len) and strings.endsWith(out, ext)) {
-            out = out[0 .. out.len - ext.len];
+            // Translated from the following JS code:
+            //   return StringPrototypeSlice(cwd, StringPrototypeIndexOf(cwd, '/'));
+            const index = std.mem.indexOfScalar(u8, normalizedCwd, CHAR_FORWARD_SLASH);
+            // Account for the -1 case of String#slice in JS land
+            return if (index != null) normalizedCwd[index.? ..len] else normalizedCwd[len - 1 ..len];
```

Comment:
Here this is a playing by JS land rules. Node has tests for overwriting/customizing `process.cwd()` which we'll likely support for interop and this will be a place ready for it. The thing with JS land is it's very wiggly and easy to accidentally end up supporting stuff just by JS's forgiving nature.